gtk/src/native_dialog.rs

```diff
@@ -0,0 +1,60 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use crate::prelude::*;
+use crate::NativeDialog;
+use crate::ResponseType;
+use std::cell::Cell;
+use std::future::Future;
+use std::pin::Pin;
+
+pub trait NativeDialogExtManual {
+    // rustdoc-stripper-ignore-next
+    /// Shows the dialog and returns a `Future` that resolves to the
+    /// `ResponseType` on response.
+    ///
+    /// ```no_run
+    /// use gtk::prelude::*;
+    ///
+    /// # async fn run() {
+    /// let dialog = gtk::FileChooserNativeBuilder::new()
+    ///    .title("Select a File")
+    ///    .build();
+    ///
+    /// dialog.run_future().await;
+    /// dialog.hide();
+    /// println!("Selected file: {:?}", dialog.file());
+    /// # }
+    /// ```
+    #[cfg(any(feature = "v3_20", feature = "dox"))]
+    #[cfg_attr(feature = "dox", doc(cfg(feature = "v3_20")))]
+    fn run_future<'a>(&'a self) -> Pin<Box<dyn Future<Output = ResponseType> + 'a>>;
+}
+
+impl<O: IsA<NativeDialog>> NativeDialogExtManual for O {
+    #[cfg(any(feature = "v3_20", feature = "dox"))]
+    #[cfg_attr(feature = "dox", doc(cfg(feature = "v3_20")))]
+    fn run_future<'a>(&'a self) -> Pin<Box<dyn Future<Output = ResponseType> + 'a>> {
+        Box::pin(async move {
+            let (sender, receiver) = futures_channel::oneshot::channel();
+
+            let sender = Cell::new(Some(sender));
+
+            let response_handler = self.connect_response(move |_, response_type| {
+                if let Some(m) = sender.replace(None) {
+                    let _result = m.send(response_type);
+                }
```

Comment:
Shouldn't the `dialog.destroy()` be called here instead of hiding the dialog later on?

Reply:
The `.hide()` is probably redundant because the window seems to be hidden by opening a file or canceling it.

`.destroy()` would imply that the `dialog.file()` in the example could fail, right? But you want to get information out of the dialog after response.